File: shared/php/query_engine.php
<?php
declare(strict_types=1);

/**
 * Simple query helpers for the PHP Level 2 tutorial.
 *
 * What this module does:
 *  - filters stored RSS entries by category
 *  - filters stored RSS entries by keyword
 *  - trims the final list to a requested limit
 */

/**
 * Filter a list of entries by keyword and/or category.
 *
 * @param array<int, array<string, string>> $entries
 * @return array<int, array<string, string>>
 */
function filterEntries(
    array $entries,
    ?string $keyword = null,
    ?string $category = null,
    ?int $limit = null
): array {
    $filteredEntries = $entries;

    if ($category !== null && $category !== '') {
        $categoryLower = strtolower($category);
        $filteredEntries = array_values(
            array_filter(
                $filteredEntries,
                static function (array $entry) use ($categoryLower): bool {
                    return strtolower((string) ($entry['category'] ?? '')) ===
                        $categoryLower;
                }
            )
        );
    }

    if ($keyword !== null && $keyword !== '') {
        $keywordLower = strtolower($keyword);
        $filteredEntries = array_values(
            array_filter(
                $filteredEntries,
                static function (array $entry) use ($keywordLower): bool {
                    $title = strtolower((string) ($entry['title'] ?? ''));
                    $summary = strtolower((string) ($entry['summary'] ?? ''));
                    return str_contains($title, $keywordLower)
                        || str_contains($summary, $keywordLower);
                }
            )
        );
    }

    if ($limit !== null) {
        return array_slice($filteredEntries, 0, $limit);
    }

    return $filteredEntries;
}

File: shared/python/query_engine.py
"""Simple query/filter helpers for RSS entries.

What this module does:
    - filters stored RSS entries by category
    - filters entries by keyword in title or summary
    - limits the final result list

Inputs:
    - a list of entry dictionaries
    - optional keyword, category, and limit values

Outputs:
    - a filtered list of entry dictionaries
"""

from __future__ import annotations

from collections.abc import Sequence


def filter_entries(
    entries: Sequence[dict[str, str]],
    keyword: str | None = None,
    category: str | None = None,
    limit: int | None = None,
) -> list[dict[str, str]]:
    """Filter entries by keyword and/or category."""
    filtered_entries = list(entries)

    if category:
        category_lower = category.lower()
        filtered_entries = [
            entry
            for entry in filtered_entries
            if str(entry.get("category", "")).lower() == category_lower
        ]

    if keyword:
        keyword_lower = keyword.lower()
        filtered_entries = [
            entry
            for entry in filtered_entries
            if keyword_lower in str(entry.get("title", "")).lower()
            or keyword_lower in str(entry.get("summary", "")).lower()
        ]

    if limit is not None:
        return filtered_entries[:limit]
    return filtered_entries
